src/java_bytecode/java_object_factory.cpp

```diff
@@ -347,6 +366,228 @@ void java_object_factoryt::gen_pointer_target_init(
+///   MUST_UPDATE_IN_PLACE: reinitialise an existing object MAY_UPDATE_IN_PLACE:
+///   generate a runtime nondet branch between the NO_ and MUST_ cases.
+void java_object_factoryt::gen_nondet_pointer_init(
+  code_blockt &assignments,
+  const exprt &expr,
+  const irep_idt &class_identifier,
+  bool create_dynamic_objects,
+  const pointer_typet &pointer_type,
+  const update_in_placet &update_in_place)
+{
+  const typet &subtype=ns.follow(pointer_type.subtype());
+  if(subtype.id()==ID_struct)
+  {
+    const struct_typet &struct_type=to_struct_type(subtype);
+    const irep_idt &struct_tag=struct_type.get_tag();
+    // If this is a recursive type of some kind, set null.
+    if(!recursion_set.insert(struct_tag).second)
+    {
+      if(update_in_place==NO_UPDATE_IN_PLACE)
+      {
+        assignments.copy_to_operands(
+          get_null_assignment(expr, pointer_type));
+      }
+      // Otherwise leave it as it is.
+      return;
+    }
+  }
+
+  code_blockt new_object_assignments;
+  code_blockt update_in_place_assignments;
+
+  if(update_in_place!=NO_UPDATE_IN_PLACE)
+  {
+    gen_pointer_target_init(
+      update_in_place_assignments,
+      expr,
+      subtype,
+      create_dynamic_objects,
+      MUST_UPDATE_IN_PLACE);
+  }
+
+  if(update_in_place==MUST_UPDATE_IN_PLACE)
+  {
+    assignments.append(update_in_place_assignments);
+    return;
+  }
+
+  // Otherwise we need code for the allocate-new-object case:
+
+  code_blockt non_null_inst;
+  gen_pointer_target_init(
+    non_null_inst,
+    expr,
+    subtype,
+    create_dynamic_objects,
+    NO_UPDATE_IN_PLACE);
+
+  // Determine whether the pointer can be null.
+  // In particular the array field of a String should not be null.
+  bool not_null=
+    assume_non_null ||
+    ((class_identifier=="java.lang.String" ||
+      class_identifier=="java.lang.StringBuilder" ||
+      class_identifier=="java.lang.StringBuffer" ||
+      class_identifier=="java.lang.CharSequence") &&
+     subtype.id()==ID_array);
+
+  if(not_null)
+  {
+    // Add the following code to assignments:
+    // <expr> = <aoe>;
+    new_object_assignments.append(non_null_inst);
+  }
+  else
+  {
+    // if(NONDET(_Bool)
+    // {
+    //    <expr> = <null pointer>
+    // }
+    // else
+    // {
+    //    <code from recursive call to gen_nondet_init() with
+    //             tmp$<temporary_counter>>
+    // }
+    auto set_null_inst=get_null_assignment(expr, pointer_type);
+
+    code_ifthenelset null_check;
+    null_check.cond()=side_effect_expr_nondett(bool_typet());
+    null_check.then_case()=set_null_inst;
+    null_check.else_case()=non_null_inst;
+
+    new_object_assignments.add(null_check);
+  }
+
+  // Similarly to above, maybe use a conditional if both the
+  // allocate-fresh and update-in-place cases are allowed:
+  if(update_in_place==NO_UPDATE_IN_PLACE)
+  {
+    assignments.append(new_object_assignments);
+  }
+  else
+  {
+    assert(update_in_place==MAY_UPDATE_IN_PLACE);
```

Comment:
This assertion needs to become an invariant with a message as to why it is an invariant.